File: Main/src/main/java/com/redelf/commons/persistance/Facade.java
package com.redelf.commons.persistance;

import android.content.Context;

import androidx.annotation.NonNull;

import com.redelf.commons.lifecycle.InitializationWithContext;
import com.redelf.commons.lifecycle.ShutdownSynchronized;
import com.redelf.commons.lifecycle.TerminationSynchronized;

public interface Facade extends

        ShutdownSynchronized,
        TerminationSynchronized,
        InitializationWithContext

{

  <T> boolean put(String key, T value);

  <T> T get(String key);

  <T> T get(String key, T defaultValue);

  long count();

  boolean deleteAll();

  boolean delete(String key);


  boolean contains(String key);

  boolean isBuilt();

  void destroy();

  class EmptyFacade implements Facade {

    @Override public <T> boolean put(String key, T value) {
      throwValidation();
      return false;
    }

    @Override public <T> T get(String key) {
      throwValidation();
      return null;
    }

    @Override public <T> T get(String key, T defaultValue) {
      throwValidation();
      return null;
    }

    @Override public long count() {
      throwValidation();
      return 0;
    }

    @Override public boolean deleteAll() {
      throwValidation();
      return false;
    }

    @Override public boolean delete(String key) {
      throwValidation();
      return false;
    }

    @Override public boolean contains(String key) {
      throwValidation();
      return false;
    }

    @Override public boolean isBuilt() {
      return false;
    }

    @Override public void destroy() {
      throwValidation();
    }

    @Override
    public boolean shutdown() {
      return true;
    }

    @Override
    public boolean terminate() {
      return true;
    }

    @Override
    public void initialize(@NonNull Context ctx) {

      // Ignore
    }

    private void throwValidation() {
      throw new IllegalStateException("Data is not built. " +
          "Please call build() and wait the initialisation finishes.");
    }
  }
}

File: Main/src/main/java/com/redelf/commons/persistance/Facade.kt
package com.redelf.commons.persistance

import android.content.Context
import com.redelf.commons.lifecycle.InitializationWithContext
import com.redelf.commons.lifecycle.ShutdownSynchronized
import com.redelf.commons.lifecycle.TerminationSynchronized

interface Facade : ShutdownSynchronized, TerminationSynchronized, InitializationWithContext {
    fun <T> put(key: String?, value: T): Boolean

    fun <T> get(key: String?): T?

    fun <T> get(key: String?, defaultValue: T): T?

    fun count(): Long

    fun deleteAll(): Boolean

    fun delete(key: String?): Boolean


    fun contains(key: String?): Boolean

    val isBuilt: Boolean

    fun destroy()

    class EmptyFacade : Facade {
        override fun <T> put(key: String?, value: T): Boolean {
            throwValidation()
            return false
        }

        override fun <T> get(key: String?): T? {
            throwValidation()
            return null
        }

        override fun <T> get(key: String?, defaultValue: T): T? {
            throwValidation()
            return null
        }

        override fun count(): Long {
            throwValidation()
            return 0
        }

        override fun deleteAll(): Boolean {
            throwValidation()
            return false
        }

        override fun delete(key: String?): Boolean {
            throwValidation()
            return false
        }

        override fun contains(key: String?): Boolean {
            throwValidation()
            return false
        }

        override fun isBuilt(): Boolean {
            return false
        }

        override fun destroy() {
            throwValidation()
        }

        override fun shutdown(): Boolean {
            return true
        }

        override fun terminate(): Boolean {
            return true
        }

        override fun initialize(ctx: Context) {
            // Ignore
        }

        private fun throwValidation() {
            throw IllegalStateException(
                "Data is not built. " +
                        "Please call build() and wait the initialisation finishes."
            )
        }
    }
}
